## Assignement 5

In [1]:
#Import Libraries
import numpy as np
import scipy.stats as st

### Question 1.1

In [2]:
#Calculta Interest
interest = (100-98)/100
strike = 60

#Calculate the exp. payoff of the call
exp_payoff_call = 0.85*(65 - strike) + 0.15*(0)
#Discount the exp. payoff to get the price
price_call = round(exp_payoff_call/(1 + interest),2)

print("The European Call has a price of ", price_call)

The European Call has a price of  4.17


### Question 1.2
Yes you could replicate the call option with bonds and the underlying such that the payoff of the call and the portfolio are the same. Then you short this portfolio witch should have a value of 4.17 and buy the call which is trading at 2. Your expected payoff is zero but you make a riskless profit of 2.17.

### Question 1.3


In [3]:
#Calculate the exp. payoff of the call
exp_payoff_put = 0.85*(0) + 0.15*(strike - 45)
#Discount the exp. payoff to get the price
price_put = round(exp_payoff_put/(1 + interest),2)

print("The European Put has a value of ", price_put)

The European Put has a value of  2.21


In [4]:
#Present value of the strike
pv_strike = round(strike/(1 + interest),2)
print(pv_strike)

58.82


In [5]:
#Check Put Call Parity 
price_call - price_put

1.96

In [6]:
5/(1 + interest)

4.901960784313726

### Question 2.1

The red line describes a call option with strike K = 30 and maturity one year <br>
The dark blue line describes a call option with strike K = 60 and maturity one year <br>
The light blue line describes a call option with strike K = 60 and maturity 6 months year <br>

The purple line describes a put option with strike K = 30 and maturity one year <br>
The yellow line describes a put option with strike K = 60 and maturity one year <br>
The grey blue line describes a put option with strike K = 60 and maturity 6 months year <br>


The reason for that is that call options exhibit a delta between 0 and 1 (option price rises is the price of the underlying rises), whereas  put options have a delta bewteen 0 and -1 (option prices goes down as the underlying rises). 
As we approach maturity the probability of getting from out-of-the-money into in-of-the-money and vice versa gets lower. For that reason more of the in-of-the-money the money options tend to exhibit the same price movments as the underlying which leads to to a steeper slope of the delta curve. 

### Question 2

In [7]:
def black_scholes(S, K, T, r, sigma, opt_type):
    
    #S: spot price
    #K: strike price
    #T: time to maturity
    #r: interest rate
    #sigma: volatility of underlying asset
    #opt_type: type of option 
    
    d1 = (np.log(S / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    d2 = (np.log(S / K) + (r - 0.5 * sigma ** 2) * T) / (sigma * np.sqrt(T))
    
    call = (S * st.norm.cdf(d1, 0.0, 1.0) - K * np.exp(-r * T) * st.norm.cdf(d2, 0.0, 1.0))
    put = (K * np.exp(-r * T) * st.norm.cdf(-d2, 0.0, 1.0) - S * st.norm.cdf(-d1, 0.0, 1.0))
    
    delta_call = st.norm.cdf(d1, 0.0, 1.0)
    delta_put = -st.norm.cdf(-d1, 0.0, 1.0)
    
    if opt_type.upper() == 'CALL':
        return np.round(call,4), np.round(delta_call,4)
    
    if opt_type.upper() == 'PUT':
        return np.round(put,4), np.round(delta_put,4)


### Question 2.2

In [8]:
#Strategy 1
sp500_price = 2740
sp500_sigma = 0.18
interest_rate_a = 0.02

opt1a_delta = black_scholes(sp500_price, 2740, 3/12, interest_rate_a, sp500_sigma,'Call')[1]
opt2a_delta = black_scholes(sp500_price, 2300, 6/12, interest_rate_a, sp500_sigma,'Call')[1]
opt3a_delta = black_scholes(sp500_price, 2990, 9/12, interest_rate_a, sp500_sigma,'Call')[1]

print("Option 1 Delta: ", opt1a_delta)
print("Option 2 Delta: ", opt2a_delta)
print("Option 3 Delta: ", opt3a_delta)

strategya_delta = 100 * opt1a_delta + 200 * opt2a_delta + 50 * opt3a_delta
print("Strategy Delta ", strategya_delta)

Option 1 Delta:  0.54
Option 2 Delta:  0.9354
Option 3 Delta:  0.3498
Strategy Delta  258.57


In [9]:
#Strategy 2
SMI_price = 8900
SMI_sigma = 0.18
interest_rate_b = 0.02

opt1b_delta = black_scholes(SMI_price, 8990, 1/12, interest_rate_b, SMI_sigma,'Put')[1]
opt2b_delta = black_scholes(SMI_price, 8500, 6/12, interest_rate_b, SMI_sigma,'Call')[1]
opt3b_delta = black_scholes(SMI_price, 8300, 12/12, interest_rate_b, SMI_sigma,'Put')[1]

print("Option 1 Delta: ", opt1b_delta)
print("Option 2 Delta: ", opt2b_delta)
print("Option 3 Delta: ", opt3b_delta)

strategyb_delta = 1 * opt1b_delta + 2 * opt2b_delta + 1 * opt3b_delta
print("Strategy Delta ", strategyb_delta)

Option 1 Delta:  -0.5539
Option 2 Delta:  0.6927
Option 3 Delta:  -0.278
Strategy Delta  0.5535


In [10]:
#Strategy 3
DAX_price = 109000
DAX_sigma = 0.12
interest_rate_c = 0.05

opt1c_delta = black_scholes(DAX_price, 109000, 6/12, interest_rate_c, DAX_sigma,'Put')[1]
opt2c_delta = black_scholes(DAX_price, 109000, 6/12, interest_rate_c, DAX_sigma,'Call')[1]


print("Option 1 Delta: ", opt1c_delta)
print("Option 2 Delta: ", opt2c_delta)

strategyc_delta = 1 * opt1c_delta + 1 * opt2c_delta
print("Strategy Delta ", strategyc_delta)

Option 1 Delta:  -0.368
Option 2 Delta:  0.632
Strategy Delta  0.264


### Question 3

In [11]:
#Parameters
interest = 0
strike = 2000
S_0 = 2000
vol = 0.2

In [12]:
def add_tree_layer(input_value):
    #Accoring to the Cox Ross Rubinstein Model
    
    h = 1/6
    
    results = []
    
    #Up
    [results.append(values * (np.exp(vol*np.sqrt(h)))) for values in input_value]
    
    #Down
    [results.append(values * (np.exp(-vol*np.sqrt(h)))) for values in input_value]
    
    return list(np.unique(np.round(results,2)))

In [13]:
tree_data = []
#Period 0
tree_data.append(S_0)
#Period 1
tree_data.append([S_0*(1 + vol), S_0 * (1 - vol)])

for i in range(1,6):
    tree_data.append(add_tree_layer(tree_data[i]))

In [14]:
[print(tree_data[i]) for i in range(1,6)] 

[2400.0, 1600.0]
[1474.55, 1736.12, 2211.83, 2604.18]
[1358.94, 1600.0, 1883.82, 2038.41, 2400.0, 2825.73]
[1252.39, 1474.55, 1736.12, 1878.59, 2044.09, 2211.83, 2604.18, 3066.13]
[1154.2, 1358.94, 1600.0, 1731.3, 1883.82, 2038.41, 2217.99, 2400.0, 2825.73, 3326.98]


[None, None, None, None, None]